App Banner: Display before Editor Welcome Tour

escapemanuele opened the pull request:
#### Changes proposed in this Pull Request

Right now, the Editor Welcome Tour is showed before the App Banner, and the latter is showed only if the WT has been dismissed and the website is older than 30 minutes.

Now, we want to show the App Banner **before** the Editor Welcome Tour. We can show the Welcome Tour immediately after dismissing the App Banner.

We also want to show the Mobile app banner only at the next visit after the blogger modal has been displayed: pdgK6S-ku-p2

Final priority list:

- blogger modal (if the user is in the Blogger flow)
- app banner (at the next visit to the Editor)
- editor welcome tour (immediately after dismissing the app banner)

#### Testing instructions

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

1.  Start Calypso, sync ETK (PCYsg-ly5-p2) and sync WBE (PCYsg-l4k-p2)
2. Test the blogger flow, (if you cannot go through the blogger flow, add 'wpcom_signup_complete_show_draft_post_modal' with value 1 in ì browser console -> Application -> Session Storage). So the order of modals will be: blogger modal -> Dismiss -> Refresh -> App Banner -> Dismiss -> Editor Welcome Tour)
 ![Screenshot 2021-12-01 at 14 48 57](https://user-images.githubusercontent.com/52076348/144246007-1d7b8a65-63b3-473f-87bc-0fafdc858771.png)
3. Test that the Blogger Modal is shown. When you dismiss that and refresh you should see the App Banner and immediately after you dismiss that, the Editor Welcome Tour.

![Screen Capture on 2021-12-02 at 11-08-31](https://user-images.githubusercontent.com/52076348/144401545-b2b0cf4d-707f-46a5-ab91-eb2eebbe96f5.gif)

https://d.pr/i/GzEjPr


Related to #58274
Fixes #58274

Commit:
App Banner showing immediately.
Welcome Tour showing only if AB not present.

At the moment only after a refresh

Commit:
Memoize isAppBannerVisible selector
Also add some tests

Commit:
Removed window.calypsoifyGutenberg.isAppBannerVisible
Used new welcome guide reducer to track if app banner gets dismissed

Commit:
Do not show App Banner is Draft Post Modal is showing (blogger flow)

Commit:
test not memoized selector

Commit:
Fixed badly formatted comment

Commit:
Trigger for tests

chriskmnds commented on apps/editing-toolkit/editing-toolkit-plugin/wpcom-block-editor-nux/src/store.js:
```diff
@@ -72,12 +72,22 @@ const shouldShowFirstPostPublishedModalReducer = ( state = false, action ) => {
 	tourRating: tourRatingReducer,
 	welcomeGuideVariant: welcomeGuideVariantReducer,
 	shouldShowFirstPostPublishedModal: shouldShowFirstPostPublishedModalReducer,
+	welcomeGuideEnabled: welcomeGuideEnabledReducer,
```
Why do we need an additional reducer here? Is this something we cannot achieve through `showWelcomeGuide` above?

escapemanuele replied:
I feel like they may refer to two separate things, with `welcomeGuideEnabled` more focused on the dismissing of the App Banner, while `showWelcomeGuide` more on if the Welcome Tour should appear or not, based on feature/user preferences.
Maybe there could be a better name for `welcomeGuideEnabled` 🤔

escapemanuele replied:
Ok @chriskmnds, that should be solved in 0e0182c :)

chriskmnds replied:
Nice! 💪

chriskmnds commented on client/blocks/app-banner/index.jsx:
```diff
@@ -161,18 +128,18 @@ export class AppBanner extends Component {
-	render() {
-		const { translate, currentSection, fetchingPreferences } = this.props;
+	isDraftPostModalShowing() {
+		return window.sessionStorage.getItem( 'wpcom_signup_complete_show_draft_post_modal' );
```
Should we guard for undefined window here, in case any test or server-side render makes a call through? (also done on line 60 above)

chriskmnds commented on client/gutenberg/editor/calypsoify-iframe.tsx:
```diff
@@ -852,6 +882,8 @@ const mapStateToProps = (
 		site: getSite( state, siteId ),
 		parentPostId,
+		isAppBannerVisible: isAppBannerVisible( state ),
+		appBannerDismissed: state.ui.appBannerDismissed,
```
A selector here would make sense. It's best I think if we avoid accessing the state tree directly where possible. Selectors being our main access layer, then state tree can also evolve independently.

chriskmnds commented on client/state/selectors/test/is-app-banner-visible.js:
```diff
@@ -0,0 +1,77 @@
+import { isAppBannerVisible } from 'calypso/state/selectors/is-app-banner-visible';
+
+describe( 'isAppBannerVisible()', () => {
+	test( 'should return false if Terms or Service update banner is displayed', () => {
```
> Terms or Service

typo :D

but maybe just ToS would suffice?

escapemanuele replied:
Ops 😄 
In latest refactor, `isAppBannerVisible` became `shouldDisplayAppBanner` so to not overlap with `isAppBannerVisible` and `isAppBannerDismissed` from state.ui

chriskmnds commented on client/state/selectors/test/is-app-banner-visible.js:
```diff
@@ -0,0 +1,77 @@
+		expect( output ).to.be.false;
+	} );
+
+	test( 'should return false is fetching preferences', () => {
```
```suggestion
	test( 'should return false if fetching preferences', () => {
```

chriskmnds reviewed (commented):
Looking good. Left a run-by review with some comments. will come back to test a little later :-)

Commit:
Refactor:
- selector for ui.appBannerDismissed
- isAppBannerVisible is shouldDisplayAppBanner
- guard for window undefined

Commit:
Typo in tests

Commit:
Removed WelcomeGuideEnabled

chriskmnds commented on apps/editing-toolkit/editing-toolkit-plugin/wpcom-block-editor-nux/src/welcome-tour/tour-launch.js:
```diff
@@ -17,15 +17,16 @@ import getTourSteps from './tour-steps';
-		isManuallyOpened: select( 'automattic/wpcom-welcome-guide' ).isWelcomeGuideManuallyOpened(),
-	} ) );
-
+	const { show, isNewPageLayoutModalOpen, isManuallyOpened, isWelcomeGuideEnabled } = useSelect(
```
```suggestion
	const { show, isNewPageLayoutModalOpen, isManuallyOpened } = useSelect(
```

chriskmnds commented on apps/editing-toolkit/editing-toolkit-plugin/wpcom-block-editor-nux/src/welcome-tour/tour-launch.js:
```diff
@@ -40,9 +41,9 @@ function LaunchWpcomWelcomeTour() {
 			is_manually_opened: isManuallyOpened,
 		} );
-	}, [ isNewPageLayoutModalOpen, isManuallyOpened, show ] );
+	}, [ isNewPageLayoutModalOpen, isManuallyOpened, show, isWelcomeGuideEnabled ] );
```
```suggestion
	}, [ isNewPageLayoutModalOpen, isManuallyOpened, show ] );
```

chriskmnds commented on apps/editing-toolkit/editing-toolkit-plugin/wpcom-block-editor-nux/src/welcome-tour/tour-launch.js:
```diff
@@ -40,9 +41,9 @@ function LaunchWpcomWelcomeTour() {
+	}, [ isNewPageLayoutModalOpen, isManuallyOpened, show, isWelcomeGuideEnabled ] );
 
-	if ( ! show || isNewPageLayoutModalOpen ) {
+	if ( ! show && ! isNewPageLayoutModalOpen ) {
```
Do we need to change the condition here?

escapemanuele replied:
Oh yes, you're right, thanks for noticing! (We don't need to change that) 😄

chriskmnds commented on apps/wpcom-block-editor/src/calypso/features/iframe-bridge-server.js:
```diff
@@ -1081,6 +1081,37 @@ function handleInlineHelpButton( calypsoPort ) {
+	);
+	port1.onmessage = ( { data } ) => {
+		const { isAppBannerVisible, hasAppBannerBeenDismissed } = data;
+		if ( hasAppBannerBeenDismissed ) {
+			dispatch( 'automattic/wpcom-welcome-guide' ).setShowWelcomeGuide( true );
+		} else if ( isAppBannerVisible ) {
+			dispatch( 'automattic/wpcom-welcome-guide' ).setShowWelcomeGuide( false );
+		}
```
Just noting that the action `setShowWelcomeGuide` does an API request to save the status of the tour (so it persists across sessions). I think this is fine for every case, but maybe confirm anyways that this is what we want 😃

escapemanuele replied:
The only problem that I see is:
- app banner is visible
- show is set to false, and status is saved
- user closes the page, without dismissing the app banner
- user reopens -> show is set to false
- app banner is dismissed -> no welcome tour showed

It could be worth having another action that doesn't do an API request!

escapemanuele replied:
This and the others should be solved in 4396643

chriskmnds commented on apps/wpcom-block-editor/src/calypso/features/iframe-bridge-server.js:
```diff
@@ -1081,6 +1081,37 @@ function handleInlineHelpButton( calypsoPort ) {
 	);
 }
 
+/**
+ * If WelcomeTour is set to show, check if the App Banner is visible.
+ * If App Banner is visible, we set the Welcome Tour to not show.
+ * When the App Banner gets dismissed, we set the Welcome Tour to show.
+ *
+ * @param {MessagePort} calypsoPort Port used for communication with parent frame.
+ */
+function getIsAppBannerVisible( calypsoPort ) {
```
I wonder if a different name here would work a little better. It might just be me, but it feels like the description (and hence logic) and the name `getIsAppBannerVisible` don't really serve each other well. We are effectively "handling" something here right? I wonder if this is like a naming convention that propagated (maybe for wrong reasons, cause other conventions appear in this file too?).

Obviously NIT. Mostly curious 😃

escapemanuele replied:
I think you're right. It propagated because, _in_ the beginning, I was using the `calypsoifyGutenberg` global variable, and so I followed what `getGutenboardingStatus` was doing. In the current approach we are "handling", as you said, the App Banner dismissal, so I will name it better 👍

chriskmnds reviewed (commented):
Thanks for the follow-up! Just a couple of comments :-)
Sorry I haven't got around to testing it yet.

Commit:
SetShowWelcomeGuide only local

escapemanuele committed:
Refactored away shouldDisplayTosUpdateBanner

simison commented:
FYI @yaelirub @annchichi @enejb this change would likely show up in metrics on your side.

simison commented:
FYI, important to confirm this will be compatible with new overlay color described in wpbArwn-3q4-p2

<img src="https://user-images.githubusercontent.com/87168/149361282-6c9f21c9-d744-411d-b008-e97a6ccb7437.png" width="320"/>

alshakero commented on client/blocks/app-banner/index.jsx:
```diff
@@ -258,15 +220,13 @@ export function buildDeepLinkFragment( currentRoute, currentSection ) {
 const mapStateToProps = ( state ) => {
 	const sectionName = getSectionName( state );
 	const isNotesOpen = isNotificationsOpen( state );
-	const currentRoute = getCurrentRoute( state );
```
This line removal seems to have broken things. `currentRoute` is undefined now.

escapemanuele replied:
Thank you for noticing that, it was left out after the rebase

alshakero reviewed (changes requested):
Found a bug.

escapemanuele committed:
hasReceivedRemotePreferences

alshakero reviewed (approved):
All works as advertised 🚀

jsnajdr commented:
Hi @escapemanuele :wave: This PR probably broke the e2e mobile test suite in `trunk`. Several e2e tests are failing because there is the App Banner covering the UI that is supposed to be tested.

Maybe your refactoring accidentally removed some "don't show the banner in e2e tests" check?

jsnajdr commented on client/state/selectors/should-display-app-banner.ts:
```diff
@@ -0,0 +1,68 @@
+		window.sessionStorage.getItem( 'wpcom_signup_complete_show_draft_post_modal' )
+	) {
+		return false;
+	}
```
The `shouldDisplayAppBanner` selector is called in the main `Layout` component every time it's re-rendered, which is very often. Reading from local storage on every Redux state change is very inefficient. Selectors should only read from `state`, not from any other data source.

escapemanuele replied:
Thanks, @jsnajdr, that is a valid point. I opened [a new issue](https://github.com/Automattic/wp-calypso/issues/60129) to solve this.

alshakero commented:
Note that tests were fixed in https://github.com/Automattic/wp-calypso/pull/60125. Thanks @escapemanuele!
